src/bvals/comms/bnd_info.cpp

```diff
@@ -36,202 +37,100 @@
-    }
-  } else if (nb.ni.ox2 > 0) {
-    sj = cellbounds.je(interior) + 1;
-    ej = cellbounds.je(interior) + Globals::nghost;
-  } else {
-    sj = cellbounds.js(interior) - Globals::nghost;
-    ej = cellbounds.js(interior) - 1;
-  }
-
-  if (nb.ni.ox3 == 0) {
-    sk = cellbounds.ks(interior);
-    ek = cellbounds.ke(interior);
-    if (pmb->block_size.nx3 > 1) {
-      if (nb.ni.ox1 != 0 && nb.ni.ox2 != 0) {
-        if (nb.ni.fi1 == 1)
-          sk += pmb->block_size.nx3 / 2;
-        else
-          ek -= pmb->block_size.nx3 / 2;
-      } else {
-        if (nb.ni.fi2 == 1)
-          sk += pmb->block_size.nx3 / 2;
-        else
-          ek -= pmb->block_size.nx3 / 2;
+  std::array<IndexRange, 3> bounds{shape.GetBoundsI(interior, el),
+                                   shape.GetBoundsJ(interior, el),
+                                   shape.GetBoundsK(interior, el)};
+
+  // Account for the fact that the neighbor block may duplicate
+  // some active zones on the loading block for face, edge, and nodal
+  // fields, so the boundary of the neighbor block is one deeper into
+  // the current block in some cases
+  std::array<int, 3> top_offset{TopologicalOffsetI(el), TopologicalOffsetJ(el),
+                                TopologicalOffsetK(el)};
+  std::array<int, 3> block_offset{ni.ox1, ni.ox2, ni.ox3};
+  std::array<int, 2> face_offset{ni.fi1, ni.fi2};
+
+  int off_idx = 0;
+  std::array<int, 3> s, e;
+  for (int dir = 0; dir < 3; ++dir) {
+    if (block_offset[dir] == 0) {
+      s[dir] = bounds[dir].s;
+      e[dir] = bounds[dir].e;
+      if (c2f &&
+          bounds[dir].e > bounds[dir].s) { // Check that this dimension has ghost zones
+        // We are sending from a coarser level to the coarse buffer of a finer level,
+        // so we need to only send the approximately half of the indices that overlap
+        // with the other block. We also send nghost "extra" zones in the interior
+        // to ensure there is enough information for prolongation. Also note for
+        // non-cell centered values the number of grid points may be odd, so we
+        // pick up an extra zone that is communicated. I think this is ok, but
+        // something to keep in mind if there are issues.
+        const int half_grid = (bounds[dir].e - bounds[dir].s + 1) / 2;
+        s[dir] += face_offset[off_idx] == 1 ? half_grid - Globals::nghost : 0;
+        e[dir] -= face_offset[off_idx] == 0 ? half_grid - Globals::nghost : 0;
       }
+      ++off_idx; // Offsets are listed in X1,X2,X3 order, should never try to access
+                 // with off_idx > 1 since all neighbors must have a non-zero block
+                 // offset in some direction
+    } else if (block_offset[dir] > 0) {
+      s[dir] = bounds[dir].e - Globals::nghost + 1 - top_offset[dir];
+      e[dir] = bounds[dir].e - top_offset[dir];
+    } else {
+      s[dir] = bounds[dir].s + top_offset[dir];
+      e[dir] = bounds[dir].s + Globals::nghost - 1 + top_offset[dir];
```

Comment:
I haven't fully reasoned out this new machinery... but it seems like a significant cleanup IMO

Reply:
I am pretty sure it is giving the same index ranges as the old code (I ran the test suite calling both and made sure the bounds were always equal), but if you get the chance it might be useful for someone else to go through it and make sure they don't see some edge case I missed.

Reply:
yeah I'll try to do that. I've been out sick the last while.